tests/tools/s3/utils.py

```diff
@@ -0,0 +1,124 @@
+
+import boto3
+import botocore.client
+
+
+def resource_has_tags(resource, tags={}):
+    """Checks if a resource defines all `tags`."""
+    return all(
+        tag_key in resource.keys() and resource[tag_key] == tag_value
+        for (tag_key, tag_value) in tags.items()
+    )
+
+
+class S3ResourceFetcher:
+    """A class for fetching vmm-reference test resources from S3."""
+
+    def __init__(
+            self,
+            resource_bucket,
+            resource_manifest_path,
+            download_location=None,
+    ):
+        """Initializes the S3 client, manifest of test resources and S3 bucket name."""
+        with open(resource_manifest_path) as json_file:
+            self._resource_manifest = json.load(json_file)
+        self._resource_bucket = resource_bucket
+        self._s3 = boto3.client(
+            's3',
+            config=botocore.client.Config(signature_version=botocore.UNSIGNED)
+        )
+        if download_location is None:
+            path = os.path.join(
+                os.path.dirname(__file__),
+                "../../../resources")
+            self._default_download_location = os.path.abspath(path)
+        else:
+            self._default_download_location = download_location
+        self.log = logging.getLogger(__name__)
+        self.log.info(
+            "Setting default download location: ",
+            self._default_download_location
+        )
+
+    def download(
+            self,
+            resource_type,
+            resource_name=None,
+            tags={},
+            version=None,
+            download_location=None,
+            first=False
+    ):
+        """Downloads **all** the resources that match the parameters.
+
+        The default version used when downloading is the latest version as defined in the
+        resource manifest.
+
+        If only one resource that confirms to the specification (i.e. type & tags) is needed,
+        then the user needs to set `first` to True. With this option, the first resource
+        that matches the parameters is downloaded.
+        """
+        version = version or self.get_latest_version()
+        download_location = download_location or self._default_download_location
+
+        if version in self._resource_manifest.keys():
+            resource_parent = self._resource_manifest[version]
+        else:
+            raise Exception("Invalid version: {}".format(version))
+
+        resources = [r for r in resource_parent
+                     if r["resource_type"] == resource_type
+                     and (resource_name is None or r["resource_name"] == resource_name)
+                     and resource_has_tags(r, tags)]
+        if len(resources) == 0:
+            self.log.error("No resources found")
+            exit(1)
+
+        if first:
+            # When only one resource is needed, we don't need to download all
+            # of them, so we make `resources` a single element array.s
+            resources = resources[:1]
```

Comment:
Why not simply return `resources[0]`. If a caller says `first` it's unlikely they are expecting a list.

Reply:
This is needed so that we download exactly one resource from S3. If we don't do this check, then more resources than are needed are going to be downloaded, but then we only return one of them.

Reply:
Oh !!! I am sorry, my bad! I was thinking this was `return`ing resources.

Reply:
No problem. I'll resolve this.